projects/pixel_contrast_cross_entropy_loss/hrnetconstrast_head.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+"""Modified from https://github.com/PaddlePaddle/PaddleSeg/
+blob/2c8c35a8949fef74599f5ec557d340a14415f20d/
+paddleseg/models/hrnet_contrast.py(Apache-2.0 License)"""
+
+from typing import List, Tuple
+
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+from torch import Tensor
+
+from mmseg.models.decode_heads.decode_head import BaseDecodeHead
+from mmseg.models.losses import accuracy
+from mmseg.registry import MODELS
+from mmseg.utils import ConfigType, SampleList
+
+
+class ProjectionHead(nn.Module):
+    """The projection head used by contrast learning.
+
+    Args:
+        dim_in (int):
+            The dimensions of input features.
+        proj_dim (int, optional):
+            The output dimensions of projection head. Default: 256.
+        proj (str, optional): The type of projection head,
+            only support 'linear' and 'convmlp'. Default: 'convmlp'.
+    """
+
+    def __init__(self, in_channels: int, out_channels=256, proj='convmlp'):
+        super().__init__()
+        if proj == 'linear':
+            self.proj = nn.Conv2d(in_channels, out_channels, kernel_size=1)
+        elif proj == 'convmlp':
+            self.proj = nn.Sequential(
+                ConvModule(in_channels, in_channels, kernel_size=1),
+                nn.Conv2d(in_channels, out_channels, kernel_size=1),
+            )
+        else:
+            raise KeyError("The type of project head only support 'linear' \
+                        and 'convmlp', but got {}.".format(proj))
+
+    def forward(self, x: Tensor) -> Tensor:
+        return F.normalize(self.proj(x), p=2.0, dim=1)
+
+
+class SegmentationHead(nn.Module):
+    """The segmentation head used by contrast learning.
+
+    Args:
+        dim_in (int):
+            The dimensions of input features.
+        proj_dim (int, optional):
+            The output dimensions of projection head. Default: 256.
+        proj (str, optional):
+            The type of projection head,
+            only support 'linear' and 'convmlp'. Default: 'convmlp'.
+    """
+
+    def __init__(self, in_channels: int, out_channels=19, drop_prob=0.1):
+        super().__init__()
+
+        self.seg = nn.Sequential(
+            ConvModule(in_channels, in_channels, kernel_size=1),
+            nn.Dropout2d(drop_prob),
+            nn.Conv2d(in_channels, out_channels, kernel_size=1),
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.seg(x)
+
+
+@MODELS.register_module()
+class ContrastHead(BaseDecodeHead):
```

Comment:
The `conv_seg` module in `BaseDecodeHead` is not used and it might cause a distributed training error. It suggests to inherit from `nn.Module`.